clippy_lints/src/missing_iterator_fold.rs

```diff
@@ -0,0 +1,83 @@
+use clippy_utils::diagnostics::span_lint;
+use rustc_hir::{AssocItemKind, Impl, Item, ItemKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for `Iterator` implementations not specializing `fold`.
+    ///
+    /// ### Why is this bad?
+    /// Specialize `fold` might provide more performant internal iteration.
+    /// Methods relying on `fold` would then be faster as well.
+    /// By default, methods that consume the entire iterator rely on it such as `for_each`, `count`...
+    ///
+    /// ### Example
+    /// ```no_run
+    /// struct MyIter(u32);
+    ///
+    /// impl Iterator for MyIter {
+    ///     type Item = u32;
+    ///     fn next(&mut self) -> Option<Self::Item> {
+    ///         # todo!()
+    ///         // ...
+    ///     }
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// struct MyIter(u32);
+    ///
+    /// impl Iterator for MyIter {
+    ///     type Item = u32;
+    ///     fn next(&mut self) -> Option<Self::Item> {
+    ///         # todo!()
+    ///         // ...
+    ///     }
+    ///     fn fold<B, F>(self, init: B, f: F) -> B
+    ///     where
+    ///         F: FnMut(B, Self::Item) -> B,
+    ///     {
+    ///         todo!()
+    ///     }
+    /// }
+    /// ```
+    #[clippy::version = "1.77.0"]
+    pub MISSING_ITERATOR_FOLD,
+    nursery,
```

Comment:
Hmmmm... what category do you think that this should be? I'm thinking about `perf`, but the lint may be too demanding to be put in perf

Reply:
This is definitely about `perf`ormance but there is also `pedantic` because this is "rather strict".

Reply:
Is this actually actionable in all cases? As in, can you always provide a more useful implementation than the default `fold` implementation? Because if not, I'd go as far as saying it could be a `restriction` lint. For example, std's `Range` `Iterator` implementation also does not implement `fold` and I'm not sure how/why it should

Reply:
Should I put it to `restriction`?